### ISSUE TITLE
Querying entities that has EcsChildOf and EcsIsA does not behave correctly

### ISSUE THREAD

**Issue by KenthJohan** [SanderMertens/flecs#907]
**Describe the bug**
Two entities with same components gives different result when querying.

**To Reproduce**
```c
#include <stdlib.h>
#include <stdio.h>
#include <stdint.h>
#include "flecs.h"

typedef struct
{
  int32_t x;
  int32_t y;
} Position;

void Sys(ecs_iter_t *it)
{
  Position * pos0 = ecs_field(it, Position, 1); // Shared
  Position * pos = ecs_field(it, Position, 2);
  char const * name0 = ecs_get_name(it->world, ecs_field_src(it, 1));
  printf("shared: %s %i %i\n", name0, pos0->x, pos0->y);
  for (int i = 0; i < it->count; i ++)
  {
    char const * name = ecs_get_name(it->world, it->entities[i]);
    printf("iterate: %s %i %i\n", name, pos[i].x, pos[i].y);
  }
}

int main(int argc, char *argv[])
{
  ecs_world_t *world = ecs_init();
  //https://www.flecs.dev/explorer/?remote=true
  ecs_singleton_set(world, EcsRest, {0});
  ECS_COMPONENT(world, Position);
  ECS_TAG(world, Start);

  ecs_system(world, {
      .entity = ecs_entity(world, {
          .name = "Sys",
          .add = { ecs_dependson(EcsOnUpdate) }
      }),
      .query.filter.instanced = true,
      .query.filter.terms = {
          {.id = ecs_id(Position), .inout = EcsInOut, .src.trav = EcsChildOf, .src.flags = EcsUp },
          //{.id = ecs_pair(ecs_id(Position), Start), .inout = EcsIn }  //Try this also
          {.id = ecs_id(Position), .inout = EcsIn }
      },
      .callback = Sys
  });

  ecs_entity_t base = ecs_new_entity(world, "base");
  //ecs_set_pair(world, base, Position, Start, {100, 200}); //Try this also
  ecs_set(world, base, Position, {100, 200});

  ecs_entity_t parent = ecs_new_entity(world, "parent");
  ecs_set(world, parent, Position, {3, 4});
  
  ecs_entity_t e1 = ecs_new_entity(world, "e1");
  ecs_add_pair(world, e1, EcsIsA, base);
  ecs_add_pair(world, e1, EcsChildOf, parent);

  ecs_entity_t e2 = ecs_new_entity(world, "e2");
  ecs_add_pair(world, e2, EcsIsA, base);
  ecs_add_pair(world, e2, EcsChildOf, parent);

  printf("ecs_progress:\n");
  while (ecs_progress(world, 0)) { }
}
```

Outputs:
```
shared: parent 3 4
iterate: e1 100 200
iterate: e2 3 4
shared: parent 3 4
iterate: e1 100 200
iterate: e2 3 4
...
```

Expected:
```
shared: parent 3 4
iterate: e1 100 200
iterate: e2 100 200
shared: parent 3 4
iterate: e1 100 200
iterate: e2 100 200
...
```

**Comment by SanderMertens** [comment]
This is not a bug, what happens is:
- `{.id = ecs_id(Position), .inout = EcsIn }` implicitly has `.trav = EcsIsA` (default, so prefab components get matched)
- `e1` and `e2` get matched, which both do not own `Position`
- Both fields are therefore shared (you can check this with `ecs_field_src`, it returns the prefab/parent)
- You're accessing a shared field with `pos[i]` where you should use `pos->` for shared fields
- `pos[1]` pokes in memory it shouldn't access, which happens to be the `parent` because `base` and `parent` are in the same table

**Comment by SanderMertens** [comment]
The instancing section of the query manual has more details on how the mechanism works: https://www.flecs.dev/flecs/md_docs_Queries.html#autotoc_md229

**Comment by KenthJohan** [comment]
The original bug report is not a bug.
Another bug was found when `.query.filter.instanced = false` which causes the query to not match but its fixed now by [32c8e91](https://github.com/SanderMertens/flecs/commit/32c8e91337eb6b436645de735d9df0a8776c683a).
